### ISSUE TITLE
Model Performance Using Smaller Batch Size on LibriTTS

### ISSUE THREAD

**Issue by Ming-er** [SWivid/F5-TTS#808]
### Checks

- [x] This template is only for usage issues encountered.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

**Env**
torch                         2.3.0+cu118
torchaudio                    2.3.0+cu118
transformers                  4.49.0
2 x 8 Tesla V100 32 GB
**Config**
datasets:
  name: LibriTTS_100_360_500
  batch_size_per_gpu: 9600
  batch_size_type: frame
  max_samples: 16
  num_workers: 16
optim:
  epochs: 180
  learning_rate: 7.5e-05
  num_warmup_updates: 20000
  grad_accumulation_steps: 1
  max_grad_norm: 1.0
  bnb_optimizer: false
model:
  name: F5TTS_Small

### Steps to Reproduce

Strictly follow the training scripts

### ✔️ Expected Behavior

Produce the Libritts results in the paper

### ❌ Actual Behavior

Hi, author, I notice that F5-TTS demonstrate an impressive result (about 2.7% wer on the ls-test-clean test set) even only trained on LibriTTS. So I want to reproduce it.
My setting is provided above. I have 16 gpus, and I set the frame size per batch to 9600 so in fact my effective batch size is about 1/2 (I have verified this proportion, and I notice that ZhiKang Niu, one of the developer, used the same setting on the LibriTTS set and he said he got a not bad result) of the recomended setting. However, I got a bad wer result on the provided ls-test-clean set with the number being 6%, while the sim and utmos score were well aligned with the paper. So, I wonder if the batch size has such a huge effect on the model performance.

**Comment by Ming-er** [comment]
> Hi [@Ming-er](https://github.com/Ming-er) , how many updates have you got now? For this batch size setting, 200k updates WER around 6%, 400k updates WER around 2%

Sorry, I forget to mention that. My update steps are set to 300k (180 epochs), and there are around 1.7k batches (about 35400 / 1.7k = 200 samples a batch) within an epoch.
OK, I got that. I will continue to train the model and try to align with the results.

**Comment by SWivid** [comment]
You could set the max updates to 600k, from our experience, small model on libritts about to converge at around 300/400k

**Comment by Ming-er** [comment]
> You could set the max updates to 600k, from our experience, small model on libritts about to converge at around 300/400k

OK, so the devil may be the total training updates instead of the batch size as the training recipe may apply a learning rate scheduler?

**Comment by Ming-er** [comment]
> My setting
> 
> datasets:
>   name: LibriTTS_100_360_500
>   batch_size_per_gpu: 38400  # 8 GPUs, 8 * 38400 = 307200
>   batch_size_type: frame  # "frame" or "sample"
>   max_samples: 64  # max sequences per batch if use frame-wise batch_size. we set 32 for small models, 64 for base models
>   num_workers: 16
> 
> optim:
>   epochs: 686
>   learning_rate: 7.5e-5
>   num_warmup_updates: 20000  # warmup steps
>   grad_accumulation_steps: 1  # note: updates = steps / grad_accumulation_steps
>   max_grad_norm: 1.0  # gradient clipping
>   bnb_optimizer: False  # use bnb 8bit AdamW optimizer or not
> 
> model:
>   name: F5TTS_Small
>   tokenizer: char
>   tokenizer_path: None  # if tokenizer = 'custom', define the path to the tokenizer you want to use (should be vocab.txt)
>   frac_lengths_mask: [0.7,1.0]
>   arch:
>     dim: 768
>     depth: 18
>     heads: 12
>     ff_mult: 2
>     text_dim: 512
>     conv_layers: 4
>     checkpoint_activations: False  # recompute activations and save memory for extra compute
>   mel_spec:
>     target_sample_rate: 24000
>     n_mel_channels: 100
>     hop_length: 256
>     win_length: 1024
>     n_fft: 1024
>     mel_spec_type: vocos  # 'vocos' or 'bigvgan'
>   vocoder:
>     is_local: False  # use local offline ckpt or not
>     local_path: None  # local vocoder path

@ZhikangNiu Hi, Zhikang. Sorry to bother you again, in the trainer.py file, I found that warmup_updates = (self.num_warmup_updates * self.accelerator.num_processes), so if I use 16 gpus for training, shall I set the num_warmup_updates to 100000 steps?

**Comment by SWivid** [comment]
> in the trainer.py file, I found that warmup_updates = (self.num_warmup_updates * self.accelerator.num_processes), so if I use 16 gpus for training, shall I set the num_warmup_updates to 100000 steps?

not need for that, cuz it is just for the purpose that num_warmup_updates will not change if gpu_num changed
https://github.com/SWivid/F5-TTS/blob/09b478b7d73b54f1c5e5e90f5372126e6d128a74/src/f5_tts/model/trainer.py#L299-L303

**Comment by Ming-er** [comment]
Hi, authors, Here are my settings,

```
datasets:
  name: LibriTTS_100_360_500
  batch_size_per_gpu: 19200  # 8 GPUs, 8 * 38400 = 307200
  batch_size_type: frame  # "frame" or "sample"
  max_samples: 16  # max sequences per batch if use frame-wise batch_size. we set 32 for small models, 64 for base models
  num_workers: 16

optim:
  epochs: 430
  learning_rate: 7.5e-5
  num_warmup_updates: 20000  # warmup updates
  grad_accumulation_steps: 1  # note: updates = steps / grad_accumulation_steps
  max_grad_norm: 1.0  # gradient clipping
  bnb_optimizer: False  # use bnb 8bit AdamW optimizer or not

model:
  name: F5TTS_Small
  tokenizer: char
  tokenizer_path: None  # if tokenizer = 'custom', define the path to the tokenizer you want to use (should be vocab.txt)
  arch:
    dim: 768
    depth: 18
    heads: 12
    ff_mult: 2
    text_dim: 512
    conv_layers: 4
    checkpoint_activations: False  # recompute activations and save memory for extra compute
  mel_spec:
    target_sample_rate: 24000
    n_mel_channels: 100
    hop_length: 256
    win_length: 1024
    n_fft: 1024
    mel_spec_type: vocos  # 'vocos' or 'bigvgan'
```
I have 16 gpus for training. There are about 1423 training steps in my batch, and I train it for 600k steps.
The followings are my results.
steps    wer        sim
100k  49.186	0.461
200k  11.322	0.578
300k  5.506	0.59
400k  3.828	0.592
500k  2.872	0.589
600k  2.814	0.585
I notice that the convegence speed is still slower than the official reports. Do I need to further enlarge the batch size? or are there any other suggestions for me to train the model better? Thanks ~

**Comment by SWivid** [comment]
Hi @Ming-er , you could turn on checkpointing (checkpoint_activations) which enables larger batchsize.
The results' reproducibility is 100% ensured as we have much experiments on that setting.

And also the fix with help faster converge #793 . If you are not in a hurry, could wait for an updated version in few weeks.

**Comment by Ming-er** [comment]
> My setting
> 
> datasets:
>   name: LibriTTS_100_360_500
>   batch_size_per_gpu: 38400  # 8 GPUs, 8 * 38400 = 307200
>   batch_size_type: frame  # "frame" or "sample"
>   max_samples: 64  # max sequences per batch if use frame-wise batch_size. we set 32 for small models, 64 for base models
>   num_workers: 16
> 
> optim:
>   epochs: 686
>   learning_rate: 7.5e-5
>   num_warmup_updates: 20000  # warmup steps
>   grad_accumulation_steps: 1  # note: updates = steps / grad_accumulation_steps
>   max_grad_norm: 1.0  # gradient clipping
>   bnb_optimizer: False  # use bnb 8bit AdamW optimizer or not
> 
> model:
>   name: F5TTS_Small
>   tokenizer: char
>   tokenizer_path: None  # if tokenizer = 'custom', define the path to the tokenizer you want to use (should be vocab.txt)
>   frac_lengths_mask: [0.7,1.0]
>   arch:
>     dim: 768
>     depth: 18
>     heads: 12
>     ff_mult: 2
>     text_dim: 512
>     conv_layers: 4
>     checkpoint_activations: False  # recompute activations and save memory for extra compute
>   mel_spec:
>     target_sample_rate: 24000
>     n_mel_channels: 100
>     hop_length: 256
>     win_length: 1024
>     n_fft: 1024
>     mel_spec_type: vocos  # 'vocos' or 'bigvgan'
>   vocoder:
>     is_local: False  # use local offline ckpt or not
>     local_path: None  # local vocoder path

Sorry to bother again. According to the advice, I try to fully follow the config provided by Zhikang. However, I found that when I set the max samples to 64 and the bs to 38400 and num_process to 8, the number of steps (batches) per epoch becomes 852. And the total update steps = 852 * 686 = 584.5k < 600k which is inconsistent with the reports? Or are there any mistakes I make?

**Comment by SWivid** [comment]
@Ming-er i think the setting zhikang provided is on 4 gpu (for your reference, it has wer2.68 sim0.597 at 400k; compared to your previous setting, larger lr cuz longer decay).
I suggested a 8*38400 batch size cuz it's the same setting to our report, you can adjust epoch with simple calculation for having desired max updates. You could refer to zhikang's response for a precise epoch number value to the report's expriment or you could just set a bit larger epoch to not have lr decay too fast.

**Comment by Ming-er** [comment]
> [@Ming-er](https://github.com/Ming-er) i think the setting zhikang provided is on 4 gpu (for your reference, it has wer2.68 sim0.597 at 400k; compared to your previous setting, larger lr cuz longer decay). I suggested a 8*38400 batch size cuz it's the same setting to our report, you can adjust epoch with simple calculation for having desired max updates. You could refer to zhikang's response for a precise epoch number value to the report's expriment or you could just set a bit larger epoch to not have lr decay too fast.

@SWivid OK, thank you for your clarification. So if Zhikang use 4 gpus, there may be 1704 steps in an epoch, and there are 686 epochs totally, so the max updates should be set to about 1200k steps as 1704 * 686 = 1169k to prevent the lr decays so fast. Thanks again for your patience.

**Comment by ZhikangNiu** [comment]
yes the setting is correct 发自我的 iPhone在 2025年2月28日，23:52，Yiming Li ***@***.***> 写道：﻿
  
@Ming-er i think the setting zhikang provided is on 4 gpu (for your reference, it has wer2.68 sim0.597 at 400k; compared to your previous setting, larger lr cuz longer decay). I suggested a 8*38400 batch size cuz it's the same setting to our report, you can adjust epoch with simple calculation for having desired max updates. You could refer to zhikang's response for a precise epoch number value to the report's expriment or you could just set a bit larger epoch to not have lr decay too fast.

@SWivid OK, thank you for your clarification. So if Zhikang use 4 gpus, there may be 1704 steps in an epoch, and there are 686 epochs totally, so the max updates should be set to about 1200k steps as 1704 * 686 = 1169k to prevent the lr decays so fast. Thanks again for your patience.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



Ming-er left a comment (SWivid/F5-TTS#808)

@Ming-er i think the setting zhikang provided is on 4 gpu (for your reference, it has wer2.68 sim0.597 at 400k; compared to your previous setting, larger lr cuz longer decay). I suggested a 8*38400 batch size cuz it's the same setting to our report, you can adjust epoch with simple calculation for having desired max updates. You could refer to zhikang's response for a precise epoch number value to the report's expriment or you could just set a bit larger epoch to not have lr decay too fast.

@SWivid OK, thank you for your clarification. So if Zhikang use 4 gpus, there may be 1704 steps in an epoch, and there are 686 epochs totally, so the max updates should be set to about 1200k steps as 1704 * 686 = 1169k to prevent the lr decays so fast. Thanks again for your patience.

**Comment by ZhikangNiu** [comment]
I will close this issue. if you have other questions, feel free re-open it

**Comment by kaleo996** [comment]
Sorry for bothering you after this issue is closed, but I have some questions:
 
1. This issue mentioned that the training converged slower than written in the official report, may I ask what is the official report you are referring to? I read the paper and didn't find how long it takes to converge on LibriTTS.
2. I am also training on V100, and I am using checkpointing as mentioned above to make the batch size up to 38400. It takes me about 1.6 seconds for one step of training. Is it normal or too slow?

Thank you in advance for your answer! ❤

**Comment by SWivid** [comment]
Hi @kaleo996 
1. The report mentioned in this repo is an openreview version, you can find it with google scholar search. Arxiv version will be updated in the future.
2. 1.6 sec per step seems reasonable on V100 to me, on A100, ~1.2 sec/step if I recall correctly

**Comment by rasenganai** [comment]
@SWivid, I was training Small_F5 model on libritts-500, on 8 h100, and the first epoch was 2s/it, whereas from 2nd epoch onwards it reaches 3it/s. Any particular reason for this?
